fix: remove stores from renamed/deleted store upgrades

technicallyty opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

- stores that were renamed are now properly deleted
- deleted/renamed and renamed stores are no longer added to `CommitInfo`
- deleted/renamed stores are now properly removed from rootmulti store's memory

ref: #7991 
closes: N/A

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

clevinson commented:
@robert-zaremba @aaronc can you of you have a look at this and give an ack?

marbar3778 reviewed (approved):
utACK.

For upgrades, would we want to keep renamed stores around and route to the old module if state is still present on the node for it?

robert-zaremba commented on store/rootmulti/store.go:
```diff
@@ -361,8 +369,19 @@ func (rs *Store) Commit() types.CommitID {
 
-	rs.lastCommitInfo = commitStores(version, rs.stores)
+	// remove remnants of removed stores
+	for sk := range rs.removalMap {
```
```suggestion
	for key := range rs.removalMap {
```

robert-zaremba reviewed (commented):
I'm missing a context here.
* What problem are we solving? #7991 doesn't provide a clear context either. 
* Why do we need to remember deleted stores?
* What should happen if someone wants to save a data to a deleted store? I suggest we should fail (return error). We are missing a test for that.
* What should happen when we restart an app?
* Documentation is missing.

technicallyty commented:
re: @robert-zaremba 
> * What problem are we solving? #7991 doesn't provide a clear context either.

Renamed stores were not being removed from storage. Since the original store from a rename was never added to the newStore map, once a commit happened - the commit store would just assume no changes were made to that store, so it would leave it as is.

> * Why do we need to remember deleted stores?

The stores that need to be deleted are only held temporarily in the `removalMap` until the next commit. Once a commit happens, the map is used to clear the root store of the stores in the map (ones that need to be deleted). Without that, renamed and deleted stores persist to memory, and renamed stores persist to storage, rather than being cleared.

> * What should happen if someone wants to save a data to a deleted store? I suggest we should fail (return error). We are missing a test for that.

Wouldn't you first have to acquire the key? I believe we already have errors to handle that situation, but i can look into fleshing out a more direct test for this situation though

> * What should happen when we restart an app?

not sure - do you see any issues that could arise from this?

> * Documentation is missing.

i can add some lines to cosmos-sdk/store readme

technicallyty commented:
re: @marbar3778 
> For upgrades, would we want to keep renamed stores around and route to the old module if state is still present on the node for it?

hmm could you explain this situation more? i'd assumed a rename with new module would be more of a hard fork situation

marbar3778 commented:
> hmm could you explain this situation more? i'd assumed a rename with new module would be more of a hard fork situation

oh, I was thinking in the case of rolling upgrades. In the case of a hard fork I don't think this would matter cause migration scripts would handle it.

technicallyty commented:
> oh, I was thinking in the case of rolling upgrades. In the case of a hard fork I don't think this would matter cause migration scripts would handle it.

ahh okay gotcha. are there situations where devs would inject `StoreUpgrades` outside of a migration script? maybe i'm missing the other use cases. my first thought here was to just add docs for devs to not run store upgrades outside a migration script.

anilCSE commented:
> > hmm could you explain this situation more? i'd assumed a rename with new module would be more of a hard fork situation
> 
> oh, I was thinking in the case of rolling upgrades. In the case of a hard fork I don't think this would matter cause migration scripts would handle it.

iirc on renaming a store, all the data is moved/copied over to the new store.

robert-zaremba reviewed (approved):
utACK
